File: windows/DMShot/Editor/StepGeometry.cs
using System.Windows;

namespace DMShot.Editor;

/// <summary>Single source of truth for numbered-step geometry: the badge circle
/// and the optional comment, drawn inside a small translucent speech bubble to
/// the right of the badge. The badge top-left is (a.X0, a.Y0). Mirrors
/// mac/Sources/DMShot/StepGeometry.swift.</summary>
public static class StepGeometry
{
    public const double CommentGap = 13;   // badge edge -> tail tip (tail points at the badge)
    // Single-line reference bubble height — tail metrics scale off this so they are
    // constant per font (editor + rendered bubble agree). Line height approximated
    // (fs*1.3) to stay thread-safe in the GDI export path.
    public static double CommentRefH(double fs) => fs * 1.3 + 2 * CommentPadV(fs);
    // Speech-bubble tail (Variant A2): the WHOLE left side is one wide arrow whose tip
    // juts toward the badge; the two shoulders and the tip are all rounded.
    public static double CommentTailLen(double fs) => CommentRefH(fs) * 0.47;
    public static double CommentShoulderR(double fs) => CommentRefH(fs) * 0.25;
    public static double CommentTipR(double fs) => CommentRefH(fs) * 0.20;

    public static double Diameter(Annotation a) => Math.Max(22, a.StrokeWidth * 7);

    // Match the badge number font (DrawWpf uses d*0.5) so number + comment agree.
    public static double CommentFontSize(Annotation a) => Diameter(a) * 0.5;

    // Bubble inner padding (text inset), proportional to the comment font.
    public static double CommentPadH(double fs) => fs * 0.5;
    public static double CommentPadV(double fs) => fs * 0.28;

    public static bool HasComment(Annotation a) => a.Kind == ToolKind.Step && !string.IsNullOrEmpty(a.Text);

    /// <summary>Top-left of the bubble, vertically centred on the badge using a
    /// single line's height (approximated from the font so it is thread-safe in
    /// the GDI export path and the editor + rendered bubble line up).</summary>
    public static Point BubbleOrigin(Annotation a)
    {
        double d = Diameter(a), fs = CommentFontSize(a);
        // body left = badge edge + gap + tail length (the tail tip sits `gap` from the badge)
        return new Point(a.X0 + d + CommentGap + CommentTailLen(fs), a.Y0 + d / 2 - CommentRefH(fs) / 2);
    }

    /// <summary>Top-left of the comment text (inside the bubble).</summary>
    public static Point CommentTextOrigin(Annotation a)
    {
        var o = BubbleOrigin(a);
        double fs = CommentFontSize(a);
        return new Point(o.X + CommentPadH(fs), o.Y + CommentPadV(fs));
    }

    /// <summary>Bubble rect for hit-testing (UI thread; uses WPF text measurement).</summary>
    public static Rect? BubbleRect(Annotation a)
    {
        if (!HasComment(a)) return null;
        double fs = CommentFontSize(a);
        var sz = TextLayout.Measure(a.Text, fs);
        var o = BubbleOrigin(a);
        return new Rect(o.X, o.Y, sz.Width + 2 * CommentPadH(fs), sz.Height + 2 * CommentPadV(fs));
    }
}

File: mac/Sources/DMShot/StepGeometry.swift
import AppKit
import CoreGraphics

/// Single source of truth for numbered-step geometry: the badge circle and the
/// optional comment, which is drawn inside a small translucent speech bubble
/// hanging to the right of the badge. Rendering, the inline comment editor, and
/// hit-testing all derive from these so they stay in agreement. The step anchor
/// (a.x / a.y) is the badge CENTRE.
enum StepGeometry {
    /// Gap between the badge edge and the bubble's left edge.
    static let commentGap: CGFloat = 10

    /// Badge radius in image pixels (matches the circle SceneRenderer draws).
    static func radius(for a: Annotation) -> CGFloat { a.strokeWidth * 4 + 8 }

    static func badgeRect(for a: Annotation) -> CGRect {
        let r = radius(for: a)
        return CGRect(x: a.x - r, y: a.y - r, width: r * 2, height: r * 2)
    }

    /// Comment font size — scales with the badge so number and comment match.
    static func commentFontSize(for a: Annotation) -> CGFloat { radius(for: a) }

    /// Bubble inner padding (text inset), proportional to the comment font.
    static func commentPadH(forFont fs: CGFloat) -> CGFloat { fs * 0.5 }
    static func commentPadV(forFont fs: CGFloat) -> CGFloat { fs * 0.28 }
    static func commentPadH(for a: Annotation) -> CGFloat { commentPadH(forFont: commentFontSize(for: a)) }
    static func commentPadV(for a: Annotation) -> CGFloat { commentPadV(forFont: commentFontSize(for: a)) }

    /// Top-left of the bubble. Vertical centre uses a single line's height
    /// (independent of the text content) so the live editor and the rendered
    /// bubble always line up.
    static func bubbleOrigin(for a: Annotation) -> CGPoint {
        let r = radius(for: a)
        let fs = commentFontSize(for: a)
        let bubbleH = TextLayout.size(" ", fontSize: fs).height + 2 * commentPadV(forFont: fs)
        return CGPoint(x: a.x + r + commentGap, y: a.y - bubbleH / 2)
    }

    /// Top-left of the comment text (inside the bubble).
    static func commentTextOrigin(for a: Annotation) -> CGPoint {
        let o = bubbleOrigin(for: a)
        return CGPoint(x: o.x + commentPadH(for: a), y: o.y + commentPadV(for: a))
    }

    /// Bubble bounding box, or nil when there is no comment text.
    static func bubbleRect(for a: Annotation) -> CGRect? {
        guard !a.text.isEmpty else { return nil }
        let fs = commentFontSize(for: a)
        let size = TextLayout.size(a.text, fontSize: fs)
        let o = bubbleOrigin(for: a)
        return CGRect(
            x: o.x, y: o.y,
            width: size.width + 2 * commentPadH(forFont: fs),
            height: size.height + 2 * commentPadV(forFont: fs))
    }

    /// Badge ∪ bubble — the grab area for moving a step.
    static func bounds(for a: Annotation) -> CGRect {
        let b = badgeRect(for: a)
        if let bub = bubbleRect(for: a) { return b.union(bub) }
        return b
    }
}
